### PR TITLE
Make sure time reset is correct

### PR DESCRIPTION
**Pull request by DamienIrving**

### CODE REVIEW

File: unseen/time_utils.py [hunk 1]
```diff
@@ -51,7 +52,7 @@ def temporal_aggregation(
     agg_method,
     variables,
     season=None,
-    reset_times=False,
+    _times=False,
```
Comment (DamienIrving) [inline unseen/time_utils.py]:
```suggestion
    reset_times=False,
```

**Comment by DamienIrving** [comment]
@stellema With [resample](https://docs.xarray.dev/en/stable/generated/xarray.Dataset.resample.html) you can only label 'left' or 'right' ('middle' would have been an alternative to adding the timedelta), so I think we should just stick with the timedelta hack.

I've added an assertion at the beginning to check that it's annual aggregation and one afterwards to double check that the month in all the output times matches the initialisation month, so I don't think the hack will ever fail but if it does an assertion will be triggered so we won't get ourselves into trouble.

**Comment by stellema** [comment]
We could also check if there are multiple months before adding the time delta, so we're not unnecessarily changing the times e.g., 
```
if reset_times:
        assert target_freq[0:2] == "YE", "--reset_times is for annual aggregation"
        diff = ds[time_dim].values[0] - start_time
        ds[time_dim] = ds[time_dim] - diff
        if np.unique(ds[time_dim].dt.month).size > 1:
            ds[time_dim] = ds[time_dim] + timedelta(days=1)
```

**Comment by stellema** [comment]
Actually, we should be able to correctly calculate the difference from the original times for all dimensions(i.e., without using `.values[0]`) if we calculate `start_time` before calling `reindex_forecast` and reset the times after calling `time_to_lead`.

Here is an example:
```
import numpy as np
import xarray as xr

from unseen.time_utils import (
    select_time_period,
    temporal_aggregation,
    datetime_to_cftime,
)
from unseen.array_handling import time_to_lead, reindex_forecast

# Create a dataset with dims=(lead, init lat, lon)
var = "var"
time_dim = "time"
init_dim = "init_date"
lead_dim = "lead_time"

N_INIT = 24
N_LEAD = 12
START = "2000-11-01"
init = xr.cftime_range(start=START, periods=N_INIT, freq="6MS")  # Nov & May init
lead = range(N_LEAD)
time = np.array([init.shift(i, freq="MS")[:N_INIT] for i in range(len(lead))])
lats = np.arange(-60, 60)
lons = np.arange(110, 160)

rng = np.random.default_rng(seed=0)
data = rng.random((len(lead), len(init), len(lats), len(lons)))
da = xr.DataArray(
    data, coords=[lead, init, lats, lons], dims=[lead_dim, init_dim, "lat", "lon"]
)
ds = da.to_dataset(name=var)
ds = ds.assign_coords({"time": ([lead_dim, init_dim], time)})

target_freq = "YE-JUN"
reset_times = True
start_time = ds[time_dim].copy()

ds = reindex_forecast(ds)
ds = ds.resample(time=target_freq).max(dim=time_dim, keep_attrs=True)
ds = time_to_lead(ds, target_freq[0])

if reset_times:
    diff = ds[time_dim] - start_time
    ds[time_dim] = ds[time_dim].astype(dtype="datetime64[ns]") - diff
    # ds[time_dim] = (ds[time_dim].dims, datetime_to_cftime(ds[time_dim]))
```

For some unknown reason, ds has an extra `lead_time` but the times are reset properly!

**Comment by DamienIrving** [comment]
I don't think `time_to_lead` is the answer - that only gets triggered if the input data doesn't have a time dimension (i.e. it has lead_time instead).

**Comment by stellema** [comment]
True. I was assuming that if the data has a time dimension then there is no point resetting the time to an initial date.

Here is an alternative method to ensure all of the months and days are the same:
```
if reset_times and target_freq[0] == "Y":
    month = start_time.dt.month.values[0]
    day = start_time.dt.day.values[0]
    ds[time_dim] = (
        ds[time_dim]
        .dt.strftime(f"%Y-{month:02d}-{day:02d}")
        .astype("datetime64[ns]")
    )
    ds[time_dim] = (ds[time_dim].dims, datetime_to_cftime(ds[time_dim]))
    assert np.unique(ds[time_dim].dt.month).size == 1
    assert np.unique(ds[time_dim].dt.month)[0] == month
```

**Comment by DamienIrving** [comment]
That looks like a good solution to me - much neater than the time diff hack.

Just confirming too that the CAFE data is one file per initialisation date (see [file list](https://github.com/AusClimateService/unseen-projects/blob/master/file_lists/CAFE_c5-d60-pX-f6_pr_files.txt)), so we never process more than one start month at a time. That means `assert np.unique(ds[time_dim].dt.month).size == 1` will never be a problem because we won't be processing May and Nov starts at the same time.

**Comment by stellema** [comment]
Good point, yeah I need to update the CAFE tasmax file list in [unseen-projects](https://github.com/AusClimateService/unseen-projects).  
Also, here is a simple test we can also add to `test_time_utils`:
```
@pytest.mark.parametrize("example_da_timeseries", ["numpy"], indirect=True)
def test_temporal_aggregation_reset_times(example_da_timeseries):
    """Test temporal_aggregation using reset_times."""
    data = example_da_timeseries
    ds = data.to_dataset(name="var")

    ds_resampled = temporal_aggregation(
        ds,
        target_freq="YE-DEC",
        input_freq="D",
        agg_method="max",
        variables=["var"],
        season=None,
        reset_times=True,
        min_tsteps=None,
        agg_dates=True,
        time_dim="time",
    )
    assert np.all(ds_resampled.time.dt.month.values == ds.time.dt.month[0].item())
```

**Comment by stellema** [comment]
These changes are included in #76

**Pull request closed without merging**